dsp_permissions_scripts/oap/update_iris.py

```diff
@@ -0,0 +1,119 @@
+from dsp_permissions_scripts.utils.dsp_client import DspClient
+from dsp_permissions_scripts.utils.get_logger import get_logger
+
+logger = get_logger(__name__)
+
+
+@dataclass
+class IRIUpdater(ABC):
+    iri: str
+    dsp_client: DspClient
+    err_msg: str | None = field(init=False, default=None)
+
+    @abstractmethod
+    def update_iri(self, new_scope: PermissionScope) -> None:
+        pass
+
+    @staticmethod
+    def from_string(string: str, dsp_client: DspClient) -> ResourceIRIUpdater | ValueIRIUpdater:
+        if re.search(r"^http://rdfh\.ch/[^/]{4}/[^/]{22}/values/[^/]{22}$", string):
+            return ValueIRIUpdater(string, dsp_client)
+        elif re.search(r"^http://rdfh\.ch/[^/]{4}/[^/]{22}$", string):
+            return ResourceIRIUpdater(string, dsp_client)
+        else:
+            raise InvalidIRIError(f"Could not parse IRI {string}")
+
+    def _get_res_dict(self, res_iri: str) -> dict[str, Any]:
+        return self.dsp_client.get(f"/v2/resources/{quote_plus(res_iri, safe='')}")
+
+
+@dataclass
+class ResourceIRIUpdater(IRIUpdater):
+    def update_iri(self, new_scope: PermissionScope) -> None:
+        res_dict = self._get_res_dict(self.iri)
+        try:
+            update_permissions_for_resource(
+                resource_iri=self.iri,
+                lmd=res_dict["knora-api:lastModificationDate"],
+                resource_type=res_dict["@type"],
+                context=res_dict["@context"] | {"knora-admin": KNORA_ADMIN_ONTO_NAMESPACE},
+                scope=new_scope,
+                dsp_client=self.dsp_client,
+            )
+        except ApiError as err:
+            self.err_msg = err.message
+            logger.error(self.err_msg)
+
+
+@dataclass
+class ValueIRIUpdater(IRIUpdater):
+    def update_iri(self, new_scope: PermissionScope) -> None:
+        res_iri = re.sub(r"/values/[^/]{22}$", "", self.iri)
+        res_dict = self._get_res_dict(res_iri)
+        val_oap = next((v for v in get_value_oaps(res_dict) if v.value_iri == self.iri), None)
+        if not val_oap:
+            self.err_msg = f"Could not find value {self.iri} in resource {res_dict['@id']}"
+            logger.error(self.err_msg)
+            return
+        val_oap.scope = new_scope
+        try:
+            update_permissions_for_value(
+                value=val_oap,
+                resource_type=res_dict["@type"],
+                context=res_dict["@context"] | {"knora-admin": KNORA_ADMIN_ONTO_NAMESPACE},
+                dsp_client=self.dsp_client,
+            )
+        except ApiError as err:
+            self.err_msg = err.message
+            logger.error(self.err_msg)
+
+
+def update_iris(
+    iri_file: Path,
+    new_scope: PermissionScope,
+    dsp_client: DspClient,
+) -> None:
+    iri_updaters = _initialize_iri_updaters(iri_file, dsp_client)
+    for iri in iri_updaters:
+        iri.update_iri(new_scope)
+    _tidy_up(iri_updaters, iri_file)
```

Comment:
On a higher level, this entire pattern seems counter-intuitive to me: These updaters are kind of a hybrid - the use of data classes suggests that they are data; the name suggests that they are functionality. This normally suggests that one didn't get the abstractions quite right.

I would suspect that there isn't even a need for classes representing "updaters" because they really only come down to a function each.  
And so all you really should need to do here, would be along the lines of:

```python
def update_iris(
    iri_file: Path,
    new_scope: PermissionScope,
    dsp_client: DspClient,
) -> None:
    iris = _read_iris_from_file(iri_file)  # for testability keep reading files on caller side
    failures = _update_iris(iris, new_scope, dsp_client)
    _log_result(failures) 

def _update_iris(
    iris: list[str],
    scope: PermissionScope,
    client: DspClient,
) -> list[IriUpdateIssue]: # or however you want to structure failure cases, could just be strings
    failures = []
    for iri in iris:
        match _get_type(iri):  # function that does the regex checks currently located in the factory, and returns an enum
            case IriType.Resource:
                if err := _update_resource_iri(iri, scope, client):  # returns either None or an error datatype
                    failures.append(err)
            case IriType.Value:
                if err := _update_value_iri(iri, scope, client):  # same
                    failures.append(err)
            case IriType.Unknown:
                failures.append(InvalidIriIssue(iri))  # again, depending on how you want to structure failures
```

I suspect this would result in simpler and more intuitive code.

Reply:
Arguably, the for loop could be extracted and a generator could be used:

```python
def update_iris(
    iri_file: Path,
    new_scope: PermissionScope,
    dsp_client: DspClient,
) -> None:
    iris = _read_iris_from_file(iri_file)
    failures = _update_iris(iris, new_scope, dsp_client)
    _log_result(failures) 

def _update_iris(
    iris: list[str],
    scope: PermissionScope,
    client: DspClient,
) -> Iterable[IriUpdateIssue]
    for iri in iris:
        if err := _update_iri(iri, scope, client):
            yield err

def _update_iri(
    iris: list[str],
    scope: PermissionScope,
    client: DspClient,
) -> IriUpdateIssue | None:
    match _get_type(iri):
        case IriType.Resource:
            if err := _update_resource_iri(iri, scope, client):
                return err
        case IriType.Value:
            if err := _update_value_iri(iri, scope, client):
                return err
        case IriType.Unknown:
            return InvalidIriIssue(iri)
    return None
```

(Also, the first code sample is missing the return statement)